include/arch/x86/ia32/arch.h

```diff
@@ -571,18 +571,84 @@ extern u32_t z_timer_cycle_get_32(void);
 extern struct task_state_segment _main_tss;
 #endif
 
+#ifdef CONFIG_USERSPACE
+/* We need a set of page tables for each thread in the system which runs in
+ * user mode. For each thread, we have:
+ *
+ *   - a toplevel PDPT
+ *   - a set of page directories for the memory range covered by system RAM
+ *   - a set of page tbales for the memory range covered by system RAM
+ *
+ * Directories and tables for memory ranges outside of system RAM will be
+ * shared and not thread-specific.
+ *
+ * NOTE: We are operating under the assumption that memory domain partitions
+ * will not be configured which grant permission to address ranges outside
+ * of system RAM.
+ *
+ * Each of these page tables will be programmed to reflect the memory
+ * permission policy for that thread, which will be the union of:
+ *
+ *   - The boot time memory regions (text, rodata, and so forth)
+ *   - The thread's stack buffer
+ *   - Partitions in the memory domain configuration (if a member of a
+ *     memory domain)
+ *
+ * The PDPT is fairly small singleton on x86 PAE and only requires 32-byte
+ * alignment, so we put it in thread->arch. The page directories and tables
+ * require page alignment so we put them in the stack object, using the below
+ * macros to compute how many pages we need.
+ */
+
+/* Define a range [Z_X86_PT_START, Z_X86_PT_END) which is the memory range
+ * covered by all the page tables needed for system RAM
+ */
+#define Z_X86_PT_START	((u32_t)ROUND_DOWN(DT_PHYS_RAM_ADDR, Z_X86_PT_AREA))
+#define Z_X86_PT_END	((u32_t)ROUND_UP(DT_PHYS_RAM_ADDR + \
+					 (DT_RAM_SIZE * 1024U), \
+					 Z_X86_PT_AREA))
```

Comment:
If the start address is rounded down, then it can be lower than DT_PHYS_RAM_ADDR, likewise the "END" address may be higher than the system RAM limit.  Is that not an overflow?  What am I missing here?

I think the idea here is that you want to detect the maximum bounding set of page addresses that encompass system RAM so that all addresses "can" be accessed through the page tables, but if (for some crazy reason) it's not actually page aligned naturally, and we actually try to use those pages for anything, then we'll end up out of bounds and anything can happen.

Obviously in practice all systems have page-aligned memory regions because that's what the hardware supports, and if we ever hit the case where it's not it's surely going to be because of a typo in the devicetree files.

This doesn't really sound like a useful feature.  I'd suggest turning this into a build assert that guarantees that the devicetree memory is page-aligned, or if some some reason we have to handle that on some hardware, we should be swapping UP/DOWN above so that the PT limits are always guaranteed to be in bounds.

Reply:
This isn't for page alignment. All addresses are assumed to be page aligned.
This is the bounds of memory covered by a *page table*, which is 2 megabytes.
This ensures that (for whatever reason) if say RAM starts at 1MB, and we have 4 MB of RAM, then we generate three page user page tables covering 0-2MB, 2MB-4MB, 4MB-6MB, since RAM spans from 1MB-5MB.